pkg/image/source.go

```diff
@@ -46,22 +64,82 @@ func ParseImageSpec(imageSpec string) (Source, string) {
 	return source, strings.TrimPrefix(imageSpec, candidates[0]+"://")
 }
 
-// ParseSource attempts to resolve a concrete image source selection from a user string (e.g. "docker://", "tar://",
-// "podman://", etc.)
+// ParseSource attempts to resolve a concrete image source selection from a user string (e.g. "docker://", "tar://", "podman://", etc.).
 func ParseSource(source string) Source {
 	source = strings.ToLower(source)
 	switch source {
-	case "tarball", "tar", "archive", "docker-archive":
+	case "tarball", "tar", "archive", "docker-archive", "docker-tar", "docker-tarball":
 		return DockerTarballSource
 	case "docker", "docker-daemon", "docker-engine":
 		return DockerDaemonSource
+	case "oci", "oci-directory", "oci-dir":
+		return OciDirectorySource
+	case "oci-tarball", "oci-tar", "oci-archive":
+		return OciTarballSource
 	case "podman":
 		// TODO: implement
 		return UnknownSource
 	}
 	return UnknownSource
 }
 
+// DetectSourceFromPath will distinguish between a oci-layout dir, oci-archive, and a docker-archive.
+func DetectSourceFromPath(imgPath string) Source {
+	return detectSourceFromPath(afero.NewOsFs(), imgPath)
+}
+
+// detectSourceFromPath will distinguish between a oci-layout dir, oci-archive, and a docker-archive for a given filesystem.
+func detectSourceFromPath(fs afero.Fs, imgPath string) Source {
+	pathStat, err := fs.Stat(imgPath)
+	if os.IsNotExist(err) {
+		return UnknownSource
+	} else if err != nil {
+		return UnknownSource
+	}
+
+	if pathStat.IsDir() {
+		//  check for oci-directory
+		if _, err := fs.Stat(path.Join(imgPath, "oci-layout")); !os.IsNotExist(err) {
+			return OciDirectorySource
+		}
+
+		// there are no other directory-based source formats supported
+		return UnknownSource
+	}
+
+	// assume this is an archive...
+	archive, err := fs.Open(imgPath)
+	if err != nil {
+		return UnknownSource
```

Comment:
This pattern is repeated in this function and it somehow doesn't look right, and it worries me that the actual error is being ignored in favor of returning `UnkownSource`. What is the reason to only return a `Source` type here instead of both `Source` and `error` ?

Reply:
Fair question, I started with including an error and then removed it (https://github.com/anchore/stereoscope/pull/40#discussion_r492986493). When I included it I felt it would make sense to bubble up the error so someone could figure out why detection failed, then I realized I wasn't doing that nor using it. 

This function is meant to be a "best effort" guess as to what the image type is from the given file/dir. But at the same time it's also changing from a "string parser" to additionally a "environment inspector" which is prone to runtime-related errors. I'll head back the direction of returning an error, and bubble it up to the stereoscope client (different from what I was doing earlier).

Reply:
Your explanation makes sense to me, I asked the question because it jumped at me, but this could go either way really. Thanks for clarifying!

Reply:
updated